### PR TITLE
[kernels] follow-ups on #641 and #644: dispatch boundary, ULP bound, block_v validation

### PR DESCRIPTION
**Pull request by wenxuewuhd**
Two follow-ups raised in review of #641 and #644.

## #644 — validate the vocab tile width

`tl.arange` needs a positive power of two, and `min(block_v, next_power_of_2(V))` only bounded the value from above. `block_v=0`, `3` or `1000` reached the kernel and failed inside the Triton compiler, where the message names neither the argument nor the constraint. They are now rejected with a `ValueError`, along with a non-int (`True` is an `int` subclass, so it needs its own arm) and an fp64 `logits`.

Bounding from above is also not enough on its own: a legal power of two can still be several tiles wide in the input dtype — fp32 `block_v=16384` is 64 KiB against a 32 KiB tile — which fails to compile rather than running slowly. The clamp is `min(block_v, budget, next_power_of_2(V))`, budget first.

An empty vocab gets its own check rather than riding the clamp: sglang replaces `triton.next_power_of_2` with a variant returning 1 at 0, so under sglang `V=0` ran at `BLOCK_V=1` — no error, garbage out.

Cases added for every rejected value, for valid overrides including one wider than the vocab so the clamp stays covered, and for the width that actually reaches the kernel: the launch is recorded and pinned to the dtype budget. Without that, the budget term was guarded only by the compile failure it prevents, so a budget that is wrong but still compiles passed unnoticed.

## #641 — dispatch boundary and the ULP bound

`split_qkv_rmsnorm_rope_pos_cache_half` takes the wide launch grid when `q_hidden_size <= _WIDE_GRID_MAX_Q_HIDDEN` (8192). That constant came from a compile probe, and the widest case under test was 6144, so neither side of the cutoff was exercised. Adds:

- `q_hidden_size = 8192` — the last width the wide grid takes;
- `q_hidden_size = 12288` — must fall back to the per-head grid. It raises `MLIRCompilationError` on the wide grid, so widening the cutoff fails here rather than at run time.

Both cases record the launch grids and assert the branch they claim. Agreement alone is not evidence: if the width check rejects the case, both arms fall back and match for the wrong reason, and "falls back" is a claim nothing checks. That is also why the per-head arm is identified by comparing the two launches rather than by `n_cols > 1` — with a single KV head, both arms are one column.

The two grid arms were compared with `atol=1e-6` plus `rtol=2**-7` and described as a 1 BF16 ULP bound. `assert_close` adds the two terms, so that admits much more than one ULP near zero and fewer than one near the top of the exponent range. Replaced with a count of steps between the bf16 bit patterns, mapped through sign-magnitude so it stays meaningful across zero, with non-finite values rejected rather than counted — two equal NaNs are zero steps apart and the largest finite value is one step from infinity, so both would pass a tight bound while meaning nothing.

That bound holds elementwise for Q, and for K only against the row's magnitude. RoPE cancels: `b*cos + a*sin` sums two terms of the row's magnitude into a result that can be far smaller, so one step of the operands survives whole and is many steps of the result. At `q_hidden` 8192 the worst element sums -0.2749 and +0.2906 to 0.0158, where the arms differ by 16 steps of the element and well under one step of its row.

The single seed was the weaker half: the old predicate passed on seed 0 and failed on 6 of 20 others, so the arms case runs four seeds, three of them ones it used to fail.

## Tests

Run on 910B3:

| File | Result |
| --- | --- |
| `tests/python/sgl_kernel_npu/test_argmax_softmax_prob.py` | 47 passed |
| `tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py` | 9 passed |

Each new assertion was checked against the mutation it guards. Two that the previous round would have missed: halving the dtype budget — a wrong width that still compiles and still computes the right answer — now fails on the width; and widening `_WIDE_GRID_MAX_Q_HIDDEN` to 16384 fails the 12288 case, today with `MLIRCompilationError` and, once that width ever compiles, on the recorded grid.

### CODE REVIEW

File: python/sgl_kernel_npu/sgl_kernel_npu/sample/argmax_softmax_prob.py [hunk 1]
```diff
@@ -93,6 +93,10 @@ def argmax_softmax_prob_fused(
     grid = (min(_num_cores(), B),)
     if block_v is None:
         block_v = _TILE_BYTES // logits.element_size()
+    elif block_v <= 0 or block_v & (block_v - 1):
```
Comment (rwang5203) [inline python/sgl_kernel_npu/sgl_kernel_npu/sample/argmax_softmax_prob.py:113]: Power-of-two validation still leaves one gap. FP32 `block_v=16384` reaches the documented 910B3 Unified Buffer compile failure.

One option is to cap overrides by dtype budget and test a large-vocabulary oversized power of two.

File: tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py [hunk 2]
```diff
@@ -130,6 +130,25 @@ def golden_split_qkv_rmsnorm_rope_pos_cache_half(
...
+    exponent range. Counting steps between the bf16 bit patterns is the check
+    the two grid arms actually have to pass.
+    """
+    ai = a.to(torch.bfloat16).cpu().view(torch.int16).to(torch.int32)
+    bi = b.to(torch.bfloat16).cpu().view(torch.int16).to(torch.int32)
+    # Map sign-magnitude to a monotone ordering so the step count is meaningful
+    # across zero.
+    ai = torch.where(ai < 0, torch.tensor(-(2**15), dtype=torch.int32) - ai, ai)
+    bi = torch.where(bi < 0, torch.tensor(-(2**15), dtype=torch.int32) - bi, bi)
+    steps = (ai - bi).abs()
```
Comment (rwang5203) [inline tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py]: The new ULP helper has a non-finite edge case. Equal NaNs map to zero and maximum finite to infinity as one step.

Consider rejecting non-finite values first and adding NaN and infinity controls.

File: tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py [hunk 3]
```diff
@@ -339,6 +357,32 @@ def test_grid_arms_agree_above_threshold(self):
             rope_dim=64,
         )
 
+    def test_dispatch_boundary_q_hidden_8192(self):
+        """q_hidden_size == _WIDE_GRID_MAX_Q_HIDDEN: the last width that takes
+        the wide grid. The constant was set from a compile probe, so without a
+        case here nothing turns red if it is widened past what the unified
+        buffer holds."""
+        self._run_case_grid_arms(
```
Comment (rwang5203) [inline tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py:507]: At 8192 the test can still use the fallback twice. Changing `<=` to `<` leaves every assertion green.

You could record the launch grid and assert the second call takes the wide branch.

File: python/sgl_kernel_npu/sgl_kernel_npu/sample/argmax_softmax_prob.py (hunk 1 above)
Comment (wenxuewuhd) [reply python/sgl_kernel_npu/sgl_kernel_npu/sample/argmax_softmax_prob.py:113]: Fixed: `min(block_v, budget, next_power_of_2(V))`. Power-of-two check added, cases for 0, -1, 3, 100, 1000, plus non-int — `True` is an `int` subclass and slipped through as 0. FYI it is a forward guarantee, not a present failure: outside SIMT mode `tl.arange` takes any width (`triton/language/semantic.py:659`). A case also asserts the launched `BLOCK_V`, which the argument check does not cover.

My bug: `V=0` relied on `next_power_of_2(0) == 0`, but sglang patches `next_power_of_2` to return 1 there. Under sglang the kernel ran at `BLOCK_V=1, V=0` — no error, garbage out. Its own check now.

File: tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py (hunk 2 above)
Comment (wenxuewuhd) [reply tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py]: Thx for the comments. Both right. Non-finite rejected first, with controls for equal NaNs, max finite vs infinity, equal infinities and a shape mismatch.

File: tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py (hunk 3 above)
Comment (wenxuewuhd) [reply tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py:507]: Right, vacuous against that mutation. Records the launch grids now; `<=` → `<` turns 8192 red.

Two more while there: `grids[0][1] > 1` is vacuous at one KV head, and the docstring credited this case with guarding the constant — that is the 12288 case.

File: tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py [hunk 4]
```diff
@@ -130,6 +131,136 @@ def golden_split_qkv_rmsnorm_rope_pos_cache_half(
...
+    bi = b.to(torch.bfloat16).cpu().view(torch.int16).to(torch.int32)
+    # Map sign-magnitude to a monotone ordering so the step count is meaningful
+    # across zero.
+    ai = torch.where(ai < 0, torch.tensor(-(2**15), dtype=torch.int32) - ai, ai)
+    bi = torch.where(bi < 0, torch.tensor(-(2**15), dtype=torch.int32) - bi, bi)
+    worst = int((ai - bi).abs().max())
+    assert worst <= ulps, f"worst elementwise distance {worst} bf16 ULP > {ulps}"
+
+
+def test_within_bf16_ulps_helper_controls():
```
Comment (rwang5203) [inline tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py:182]: The [attention check](https://github.com/sgl-project/sgl-kernel-npu/actions/runs/32475578357/job/96751155649) runs this file through `unittest.main()`, so both module-level pytest controls are never collected.

Move both controls into the `TestCase` or run this file with pytest.

Comment (wenxuewuhd) [reply tests/python/sgl_kernel_npu/test_split_qkv_rmsnorm_rope_pos_cache_half_npu.py]: `unittest.main()` only discovers TestCase subclasses, so both controls were dead in the attention check. Moved them into `TestUlpHelperControls` (db40cb4), outside the NPU-gated case since they touch no device — `python3 <file>` now runs 9 tests instead of 7, pytest collects the same 9. Also in 9c45cf8: `test_argmax_softmax_prob.py` and `test_verify_tree.py` both discarded `pytest.main`'s return value, so the `python3 <file>` CI branch exited 0 on failures — both now `raise SystemExit(pytest.main(...))`; the argmax run gains `-rs` so the HBM-gated skip shows its reason.
